janus_server/src/datastore.rs

```diff
@@ -1848,6 +2030,63 @@ pub mod models {
     {
     }
 
+    /// CollectJob represents a row in the `collect_jobs` table, used by leaders to represent
+    /// running collect jobs and store the results of completed ones.
+    #[derive(Clone, Derivative)]
+    #[derivative(Debug)]
+    pub(crate) struct CollectJob<A: vdaf::Aggregator>
+    where
+        for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+    {
+        /// The unique identifier for the collect job.
+        pub(crate) collect_job_id: Uuid,
+        /// The task ID for this aggregate share .
+        pub(crate) task_id: TaskId,
+        /// The batch interval covered by the aggregate share.
+        pub(crate) batch_interval: Interval,
+        /// The VDAF aggregation parameter used to prepare and aggregate input shares.
+        #[derivative(Debug = "ignore")]
+        pub(crate) _aggregation_param: A::AggregationParam,
+        /// The helper's encrypted aggregate share over the input shares in the interval. `None`
+        /// until the helper has serviced an `AggregateShareReq` for this collect job.
+        #[derivative(Debug = "ignore")]
+        pub(crate) helper_aggregate_share: Option<HpkeCiphertext>,
+        /// The leader's aggregate share over the input shares in the interval. `None` until the
+        /// leader has computed this collect job's aggregate share.
+        #[derivative(Debug = "ignore")]
+        pub(crate) leader_aggregate_share: Option<A::AggregateShare>,
+        /// The number of reports included in the aggregate share, or `None` until the leader has
+        /// computed it.
+        pub(crate) report_count: Option<u64>,
+        /// Checksum over the aggregated report shares, as described in §4.4.4.3, or `None` until
+        /// the leader has computed it.
+        #[derivative(Debug = "ignore")]
+        pub(crate) checksum: Option<[u8; 32]>,
+    }
+
+    impl<A: vdaf::Aggregator> CollectJob<A>
+    where
+        for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+    {
+        /// Returns `Ok(true)` if this collect job has already been run, `Ok(false)` if not, and an
+        /// error if the job is in an inconsistent state.
+        pub(crate) fn has_run(&self) -> Result<bool, Error> {
+            match (
+                self.helper_aggregate_share.as_ref(),
+                self.leader_aggregate_share.as_ref(),
+                self.report_count,
+                self.checksum,
+            ) {
+                (Some(_), Some(_), Some(_), Some(_)) => Ok(true),
+                (None, None, None, None) => Ok(false),
+                _ => Err(Error::DbState(format!(
+                    "collect job {} in inconsistent state",
+                    self.collect_job_id
+                ))),
```

Comment:
Depending on how we implement the rest of the leader worker processes, we may store the leader aggregate share, report count, and checksum immediately when they are computed, and then store the helper aggregate share once it has been requested and received. If so, we'd want to return `Ok(false)` for that combination. In the alternative, the leader worker would compute its fields, send a request, and only write to the database once upon receiving a successful response from the helper.

Reply:
You're right: the leader must compute its aggregate share, report count and the checksum before it sends `AggregateShareReq` to helper, and there's no reason not to record all those values in the datastore while leader waits for helper's aggregate share. Neither would we want to hold a DB txn open while the leader awaits a helper response, so we should handle the case where we have leader share, report count and checksum.